### ISSUE TITLE
MP4Box does not concatenate videos revisited

### ISSUE THREAD

**Issue by Sylvester-C** [gpac/gpac#992]
I have looked at:

https://github.com/gpac/gpac/issues/549

https://github.com/gpac/gpac/issues/235

https://github.com/gpac/gpac/issues/951

I found the discussions there interesting.

I have about 6 mp4 videos that have been encoded using FFmpeg.  All have AVC and AAC. Frame rate mode Constant.  Frame rate: 23.976 (24000/1001) FPS

I could use FFmpeg to concatenate, but with  MP4Box, I should just be able to enter MP4Box -cat video_1.mp4 -cat video_2.mp4 -new concated.mp4.  This looks nice and easy.

The problem surfaces with the program MediaInfo which provides easy access to technical and tag information about video and audio files. I have used this program for many years and find it invaluable.

Using MP4Box to concatenate my videos, the frame rate is changed to variable.

Frame rate mode     : Variable
Frame rate          : 23.976 (23976/1000) FPS
Minimum frame rate  : 15.434 FPS
Maximum frame rate  : 23.976 FPS

I have tried these option: -force-cat -add -nodrop -fps.  But they do not keep the frame rate constant.

I have also notices the audio Bit rate mode changes from constant to Variable with a Maximum bit rate and Minimum bit rate.

Is there some way I can keep constant rates please?

**Comment by rbouqueau** [comment]
Could you provide sample files please?

**Comment by Sylvester-C** [comment]
Here they are.  Just 2 of them for concatenating.  If you can join and keep constants, please let me know the options.
[film_4.zip](https://github.com/gpac/gpac/files/1743466/film_4.zip)
[film_3.zip](https://github.com/gpac/gpac/files/1743467/film_3.zip)

**Comment by rbouqueau** [comment]
Thanks. This is a duplicate of #13 so I will close it after this message.

Your both files have different properties:
 1. They do not expose the same properties (SAR 1:1). This is probably FFmpeg forwarding attributes from the source.
 2. The VBR stuff of MediaInfo comes from the fact the audio and the video tracks don't have the same duration. So when we concatenate it creates samples with a different duration to keep the a/v sync.

Traces:
```
MP4Box -info film_3.mp4
* Movie Info *
        Timescale 1000 - 2 tracks
        Computed Duration 00:00:26.444 - Indicated Duration 00:00:26.467
        Fragmented File: no
        File Brand isom - version 512
                Compatible brands: isom iso2 avc1 mp41
        Created: UNKNOWN DATE   Modified: UNKNOWN DATE
File has no MPEG4 IOD/OD

iTunes Info:
        Encoder Software: Lavf55.30.100
1 UDTA types: meta (1)

Track # 1 Info - TrackID 1 - TimeScale 24000
Media Duration 00:00:26.443 - Indicated Duration 00:00:26.443
Track has 1 edit lists: track duration is 00:00:26.444
Media Info: Language "English (eng)" - Type "vide:avc1" - 634 samples
Visual Track layout: x=0 y=0 width=1024 height=576
MPEG-4 Config: Visual Stream - ObjectTypeIndication 0x21
AVC/H264 Video - Visual Size 1024 x 576
        AVC Info: 1 SPS - 1 PPS - Profile High @ Level 3.2
        NAL Unit length bits: 32
        Pixel Aspect Ratio 1:1 - Indicated track size 1024 x 576        => NOT PRESENT IN film_4.mp4
        Chroma format YUV 4:2:0 - Luma bit depth 8 - chroma bit depth 8
        SPS#1 hash: 12C9E437ED9F9136CF7B9488BB2F1DAFE4445C7F            => DIFFERENT IN film_4.mp4
        PPS#1 hash: B98571597DA44434BD1527532A2FC01E41DA39E2            => DIFFERENT IN film_4.mp4
Self-synchronized
        RFC6381 Codec Parameters: avc1.640020
        Average GOP length: 317 samples

Track # 2 Info - TrackID 2 - TimeScale 44100
Media Duration 00:00:26.466 - Indicated Duration 00:00:26.466           => NOT THE SAME AS AUDIO
Track has 1 edit lists: track duration is 00:00:26.443
Media Info: Language "English (eng)" - Type "soun:mp4a" - 1140 samples
MPEG-4 Config: Audio Stream - ObjectTypeIndication 0x40
MPEG-4 Audio AAC LC - 2 Channel(s) - SampleRate 44100
Synchronized on stream 1
        RFC6381 Codec Parameters: mp4a.40.2
Alternate Group ID 1
        All samples are sync


>MP4Box -info film_4.mp4
* Movie Info *
        Timescale 1000 - 2 tracks
        Computed Duration 00:00:20.229 - Indicated Duration 00:00:20.252
        Fragmented File: no
        File Brand isom - version 512
                Compatible brands: isom iso2 avc1 mp41
        Created: UNKNOWN DATE   Modified: UNKNOWN DATE
File has no MPEG4 IOD/OD

iTunes Info:
        Encoder Software: Lavf55.30.100
1 UDTA types: meta (1)

Track # 1 Info - TrackID 1 - TimeScale 24000
Media Duration 00:00:20.228 - Indicated Duration 00:00:20.228
Track has 1 edit lists: track duration is 00:00:20.229
Media Info: Language "English (eng)" - Type "vide:avc1" - 485 samples
Visual Track layout: x=0 y=0 width=1024 height=576
MPEG-4 Config: Visual Stream - ObjectTypeIndication 0x21
AVC/H264 Video - Visual Size 1024 x 576
        AVC Info: 1 SPS - 1 PPS - Profile High @ Level 3.2
        NAL Unit length bits: 32
        Chroma format YUV 4:2:0 - Luma bit depth 8 - chroma bit depth 8
        SPS#1 hash: CEA0C35A05642D87E836FD53A222C3D7DF1D5555
        PPS#1 hash: A4062166489D00A67AC21A41177C26CD528FC65C
Self-synchronized
        RFC6381 Codec Parameters: avc1.640020
        Average GOP length: 485 samples

Track # 2 Info - TrackID 2 - TimeScale 44100
Media Duration 00:00:20.251 - Indicated Duration 00:00:20.251
Track has 1 edit lists: track duration is 00:00:20.228
Media Info: Language "English (eng)" - Type "soun:mp4a" - 873 samples
MPEG-4 Config: Audio Stream - ObjectTypeIndication 0x40
MPEG-4 Audio AAC LC - 2 Channel(s) - SampleRate 44100
Synchronized on stream 1
        RFC6381 Codec Parameters: mp4a.40.2
Alternate Group ID 1
        All samples are sync

```

**Comment by rbouqueau** [comment]
We can still discuss your specific use-case here if you have questions. If we find a solution to this (e.g. FFmpeg command-line) we would create a blog article so that everyone knows.

**Comment by Sylvester-C** [comment]
Thanks for looking into this and mentioning that the video and audio time have different lengths.

I changed the length of the audio to match the video time in each file exactly to the millisecond.  Concatenated again with MP4Box.  Variable results again.

Frame rate mode                          : Variable
Frame rate                               : 23.949 FPS
Minimum frame rate                       : 10.331 FPS
Maximum frame rate                       : 23.976 FPS

In fact it is worse by having a minimum frame rate at 10.331 fps.

I tried another encoder - MEncoder - on exactly the same files and then used MP4Box to join, but the frame rate shifted from constant to variable again.  I thought this would happen anyway, both programs use Lavf57.58.101 as the writing library.

In the meantime, I will go back to MEncoder AVI container method or FFmpeg MPEG-TS container method to concatenate, but will come back here from time-to-time to see if there is a fix for this issue.
